File: src/config/env.ts
import { z } from 'zod';
import { config } from 'dotenv';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Load environment variables
config();

const envSchema = z.object({
  NODE_ENV: z.enum(['development', 'production', 'test']).default('development'),
  PORT: z.string().transform(Number).default('3001'),
  LOG_LEVEL: z.enum(['error', 'warn', 'info', 'debug']).default('info'),
  DATABASE_PATH: z.string().default(path.resolve(process.cwd(), 'data/streams.db')),
  HOLODEX_API_KEY: z.string({
    required_error: 'HOLODEX_API_KEY is required'
  }),
  TWITCH_CLIENT_ID: z.string({
    required_error: 'TWITCH_CLIENT_ID is required'
  }),
  TWITCH_CLIENT_SECRET: z.string({
    required_error: 'TWITCH_CLIENT_SECRET is required'
  })
});

export const env = envSchema.parse(process.env); 

File: rust/src/config/env.rs
use serde::Deserialize;
use std::env;

#[derive(Debug, Clone, Deserialize)]
#[allow(unused)]
pub struct Env {
    pub node_env: String,
    pub port: u16,
    pub log_level: String,
    pub database_path: String,
    pub holodex_api_key: String,
    pub twitch_client_id: String,
    pub twitch_client_secret: String,
    pub youtube_api_key: Option<String>,
    pub livelink_config_dir: Option<String>,
}

impl Env {
    pub fn load() -> Self {
        dotenv::dotenv().ok();

        Self {
            node_env: env::var("NODE_ENV").unwrap_or_else(|_| "development".into()),
            port: env::var("PORT")
                .unwrap_or_else(|_| "3001".into())
                .parse()
                .unwrap_or(3001),
            log_level: env::var("LOG_LEVEL").unwrap_or_else(|_| "info".into()),
            database_path: env::var("DATABASE_PATH")
                .unwrap_or_else(|_| "data/streams.db".into()),
            holodex_api_key: env::var("HOLODEX_API_KEY").unwrap_or_default(),
            twitch_client_id: env::var("TWITCH_CLIENT_ID").unwrap_or_default(),
            twitch_client_secret: env::var("TWITCH_CLIENT_SECRET").unwrap_or_default(),
            youtube_api_key: env::var("YOUTUBE_API_KEY").ok(),
            livelink_config_dir: env::var("LIVELINK_CONFIG").ok(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_env_load_defaults() {
        let env = Env::load();
        assert!(env.port > 0);
    }
}